osin/v1/types.go

```diff
@@ -7,17 +7,56 @@ import (
 	configv1 "github.com/openshift/api/config/v1"
 )
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type OsinServerConfig struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// provides the standard apiserver configuration
+	configv1.GenericAPIServerConfig `json:",inline"`
+
+	// oauthConfig holds the necessary configuration options for OAuth authentication
+	OAuthConfig OAuthConfig `json:"oauthConfig"`
+}
+
 // OAuthConfig holds the necessary configuration options for OAuth authentication
 type OAuthConfig struct {
 	// masterCA is the CA for verifying the TLS connection back to the MasterURL.
+	// This field is deprecated and will be removed in a future release.
+	// See loginURL for details.
+	// Deprecated
 	MasterCA *string `json:"masterCA"`
 
 	// masterURL is used for making server-to-server calls to exchange authorization codes for access tokens
+	// This field is deprecated and will be removed in a future release.
+	// See loginURL for details.
+	// Deprecated
 	MasterURL string `json:"masterURL"`
 
 	// masterPublicURL is used for building valid client redirect URLs for internal and external access
+	// This field is deprecated and will be removed in a future release.
+	// See loginURL for details.
+	// Deprecated
 	MasterPublicURL string `json:"masterPublicURL"`
 
+	// loginURL, along with masterCA, masterURL and masterPublicURL have distinct
+	// meanings depending on how the OAuth server is run.  The two states are:
+	// 1. embedded in the kube api server (all 3.x releases)
+	// 2. as a standalone external process (all 4.x releases)
+	// in the embedded configuration, loginURL is equivalent to masterPublicURL
+	// and the other fields have functionality that matches their docs.
+	// in the standalone configuration, the fields are used as:
+	// loginURL is the URL required to login to the cluster:
+	// oc login --server=<loginURL>
+	// masterPublicURL is the issuer URL
+	// it is accessible from inside (service network) and outside (ingress) of the cluster
+	// masterURL is the loopback variation of the token_endpoint URL with no path component
+	// it is only accessible from inside (service network) of the cluster
+	// masterCA is used to perform TLS verification for connections made to masterURL
+	// For further details, see the IETF Draft:
+	// https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
+	LoginURL string `json:"loginURL"`
```

Comment:
Should we be keep consistent with the oauth spec and call this like `AuthorizationEndpoint`?

Reply:
This is not the `AuthorizationEndpoint` - see the `oc` command on how it is used.